models/hubble_exchange/avalanche_c/hubble_exchange_avalanche_c_perpetual_trades.sql

```diff
@@ -0,0 +1,129 @@
+{{ config(
+    alias = 'perpetual_trades',
+    partition_by = ['block_date'],
+    materialized = 'incremental',
+    file_format = 'delta',
+    incremental_strategy = 'merge',
+    unique_key = ['block_date', 'blockchain', 'project', 'version', 'tx_hash', 'evt_index'],
+    post_hook='{{ expose_spells(\'["avalanche_c"]\',
+                                "project",
+                                "hubble_exchange",
+                                \'["Henrystats"]\') }}'
+    )
+}}
+
+{% set project_start_date = '2022-08-09' %}
+
+WITH 
+
+perp_events as (
+    SELECT evt_block_time                                              as block_time,
+           evt_block_number                                            as block_number,
+           CASE WHEN (baseAsset * 1) >= 0 THEN 'long' ELSE 'short' END as trade_type,       -- negative baseAsset is for short and positive is for long
+           'AVAX'                                                      as virtual_asset,    -- only AVAX can currently be traded on hubble exchange
+           ''                                                          as underlying_asset, -- there's no way to track the underlying asset as traders need to deposit into their margin account before they're able to trade which is tracked in a seperate event not tied to the margin positions opened.
+           quoteAsset / 1E6                                            as volume_usd,
+           CAST(NULL as double)                                        as fee_usd,          -- no event to track fees
+           CAST(NULL as double)                                        as margin_usd,       -- no event to track margin
+           CAST(quoteAsset as double)                                  as volume_raw,
+           trader,
+           contract_address                                            as market_address,
+           evt_index,
+           evt_tx_hash                                                 as tx_hash
+    FROM 
+    {{ source('hubble_exchange_avalanche_c', 'ClearingHouse_evt_PositionModified') }}
+    {% if not is_incremental() %}
+    WHERE evt_block_time >= '{{project_start_date}}'
+    {% endif %}
+    {% if is_incremental() %}
+    WHERE evt_block_time >= date_trunc("day", now() - interval '1 week')
+    {% endif %}
+), 
+
+trade_data as (
+    -- close position calls 
+    SELECT call_block_number as block_number,
+           call_tx_hash      as tx_hash,
+           'close'           as trade_data
+    FROM 
+    {{ source('hubble_exchange_avalanche_c', 'ClearingHouse_call_closePosition') }}
+    WHERE call_success = true 
+    {% if not is_incremental() %}
+    AND call_block_time >= '{{project_start_date}}'
+    {% endif %}
+    {% if is_incremental() %}
+    AND call_block_time >= date_trunc("day", now() - interval '1 week')
+    {% endif %}
+
+    UNION
```

Comment:
union all? or needs distinct?

Reply:
whilst I was working on tigris spells a few weeks ago, there was a weird event where someone is able to close/open a position at the same time. -- although that was an hack 

The reason I put union there is to prevent duplicates like that. But UNION will be fine.